python/cudf/cudf/core/indexed_frame.py

```diff
@@ -756,6 +757,47 @@ def add_suffix(self, suffix):
                 Use `Series.add_suffix` or `DataFrame.add_suffix`"
         )
 
+    @annotate("APPLY", color="purple", domain="cudf_python")
+    def _apply(self, func, kernel_getter, *args, **kwargs):
+        """Apply `func` across the rows of the frame."""
+        if kwargs:
+            raise ValueError("UDFs using **kwargs are not yet supported.")
+
+        try:
+            kernel, retty = _compile_or_get(
+                self, func, args, kernel_getter=kernel_getter
+            )
+        except Exception as e:
+            raise ValueError(
+                "user defined function compilation or execution failed."
+            ) from e
+
+        # Mask and data column preallocated
+        ans_col = cp.empty(len(self), dtype=retty)
+        ans_mask = cudf.core.column.column_empty(len(self), dtype="bool")
+        launch_args = [(ans_col, ans_mask), len(self)]
+        offsets = []
+
+        # if _compile_or_get succeeds, it is safe to create a kernel that only
+        # consumes the columns that are of supported dtype
+        for col in _supported_cols_from_frame(self).values():
+            data = col.data
+            mask = col.mask
+            if mask is None:
+                launch_args.append(data)
+            else:
+                launch_args.append((data, mask))
+            offsets.append(col.offset)
+        launch_args += offsets
+        launch_args += list(args)
+        kernel.forall(len(self))(*launch_args)
```

Comment:
If we always generate a kernel with `len(self)` tasks, do we really need to pass `len(self)` as one of the `launch_args`? AFAICT that's just used to avoid out of bounds accesses, but it looks like we always launch a grid with one thread per row right?

Reply:
I had taken the numba docs for `forall` to mean that inserting this check was a requirement for kernels that expect to be configured this way. Indeed, taking it out leads to numerous tests failing due to nulls being in the wrong places everywhere. My assumption was that something was happening inside `forall` that caused unpredictable behavior if this guard was not included. cc @gmarkall for more insight.

Reply:
I guess maybe numba must be doing something where it generates only a limited set of kernels (maybe templates?) based on the block size and then dispatches to the closest size possible based on the argument to `forall`? I would be curious to learn more about how this works from @gmarkall. It sounds like you should ignore my suggestion to change anything here though.